File: i2f-jdk/i2f-swl/src/main/java/i2f/swl/test/TestSwlExchanger.java
package i2f.swl.test;

import i2f.jce.std.encrypt.asymmetric.key.AsymKeyPair;
import i2f.swl.cert.data.SwlCert;
import i2f.swl.cert.data.SwlCertPair;
import i2f.swl.core.exchanger.SwlExchanger;
import i2f.swl.data.SwlData;
import i2f.swl.impl.SwlAesSymmetricEncryptor;
import i2f.swl.impl.SwlBase64Obfuscator;
import i2f.swl.impl.SwlRsaAsymmetricEncryptor;
import i2f.swl.impl.SwlSha256MessageDigester;

import java.util.Arrays;

/**
 * @author Ice2Faith
 * @date 2024/7/10 10:12
 * @desc
 */
public class TestSwlExchanger {
    public static void main(String[] args) throws Exception {
//        testRaw();

        testCert();
    }

    public static void testCert() throws Exception {
        SwlExchanger clientTransfer = new SwlExchanger();
        clientTransfer.setAsymmetricEncryptorSupplier(() -> new SwlRsaAsymmetricEncryptor());
        clientTransfer.setSymmetricEncryptorSupplier(() -> new SwlAesSymmetricEncryptor());
        clientTransfer.setMessageDigester(new SwlSha256MessageDigester());
        clientTransfer.setObfuscator(new SwlBase64Obfuscator());

        SwlExchanger serverTransfer = new SwlExchanger();
        serverTransfer.setAsymmetricEncryptorSupplier(() -> new SwlRsaAsymmetricEncryptor());
        serverTransfer.setSymmetricEncryptorSupplier(() -> new SwlAesSymmetricEncryptor());
        serverTransfer.setMessageDigester(new SwlSha256MessageDigester());
        serverTransfer.setObfuscator(new SwlBase64Obfuscator());


        SwlCertPair certPair = clientTransfer.generateCertPair("test");
        SwlCert serverCert = certPair.getServer();
        SwlCert clientCert = certPair.getClient();


        SwlData clientSendData = clientTransfer.sendByCert(clientCert,
                Arrays.asList("body:123456", "query:user=admin"));


        String clientId = "127.0.0.1";
        SwlData serverReceiveData = serverTransfer.receiveByCert(clientId,
                clientSendData,
                serverCert);

        SwlData serverResponseData = serverTransfer.sendByCert(serverCert,
                Arrays.asList("echo:ok", "data:ok"));

        String serverId = "server";
        SwlData clientReceiveData = clientTransfer.receiveByCert(
                serverResponseData,
                clientCert
        );

        System.out.println("ok");
    }

    public static void testRaw() throws Exception {

        SwlExchanger clientTransfer = new SwlExchanger();

        clientTransfer.setAsymmetricEncryptorSupplier(() -> new SwlRsaAsymmetricEncryptor());
        clientTransfer.setSymmetricEncryptorSupplier(() -> new SwlAesSymmetricEncryptor());
        clientTransfer.setMessageDigester(new SwlSha256MessageDigester());
        clientTransfer.setObfuscator(new SwlBase64Obfuscator());

        SwlExchanger serverTransfer = new SwlExchanger();
        serverTransfer.setAsymmetricEncryptorSupplier(() -> new SwlRsaAsymmetricEncryptor());
        serverTransfer.setSymmetricEncryptorSupplier(() -> new SwlAesSymmetricEncryptor());
        serverTransfer.setMessageDigester(new SwlSha256MessageDigester());
        serverTransfer.setObfuscator(new SwlBase64Obfuscator());


        AsymKeyPair serverKeyPair = clientTransfer.generateKeyPair();
        String serverAsymSign = "server";

        AsymKeyPair clientKeyPair = clientTransfer.generateKeyPair();
        String clientAsymSign = "client";


        SwlData clientSendData = clientTransfer.sendByRaw(serverKeyPair.getPublicKey(), serverAsymSign,
                clientKeyPair.getPrivateKey(), clientAsymSign,
                Arrays.asList("body:123456", "query:user=admin"));


        String clientId = "127.0.0.1";
        SwlData serverReceiveData = serverTransfer.receiveByRaw(clientId,
                clientSendData,
                clientKeyPair.getPublicKey(),
                serverKeyPair.getPrivateKey());

        SwlData serverResponseData = serverTransfer.sendByRaw(clientKeyPair.getPublicKey(), clientAsymSign,
                serverKeyPair.getPrivateKey(), serverAsymSign
                , Arrays.asList("echo:ok", "data:ok"));

        String serverId = "server";
        SwlData clientReceiveData = clientTransfer.receiveByRaw(serverId,
                serverResponseData,
                serverKeyPair.getPublicKey(),
                clientKeyPair.getPrivateKey());

        System.out.println("ok");
    }
}

File: i2f-jdk/i2f-swl/web/i2f-swl/src/i2f-turbo-web/i2f-swl/test/TestSwlExchanger.js
import SwlTransfer from "../core/core/SwlTransfer";
import SwlRsaAsymmetricEncryptorSupplier from "../core/impl/supplier/SwlRsaAsymmetricEncryptorSupplier";
import SwlAesSymmetricEncryptorSupplier from "../core/impl/supplier/SwlAesSymmetricEncryptorSupplier";
import SwlSha256MessageDigester from "../core/impl/SwlSha256MessageDigester";
import SwlBase64Obfuscator from "../core/impl/SwlBase64Obfuscator";
import SwlExchanger from "@/i2f-turbo-web/i2f-swl/core/core/exchanger/SwlExchanger";

/**
 * @return TestSwlExchanger
 * @constructor
 */
function TestSwlExchanger() {

}

TestSwlExchanger.main=function(){
    // TestSwlExchanger.testRaw();

    TestSwlExchanger.testCert();
}

TestSwlExchanger.testCert = function () {

    let clientTransfer = new SwlExchanger();
    clientTransfer.asymmetricEncryptorSupplier = (new SwlRsaAsymmetricEncryptorSupplier());
    clientTransfer.symmetricEncryptorSupplier = (new SwlAesSymmetricEncryptorSupplier());
    clientTransfer.messageDigester = (new SwlSha256MessageDigester());
    clientTransfer.obfuscator = (new SwlBase64Obfuscator());

    let serverTransfer = new SwlExchanger();
    serverTransfer.asymmetricEncryptorSupplier = (new SwlRsaAsymmetricEncryptorSupplier());
    serverTransfer.symmetricEncryptorSupplier = (new SwlAesSymmetricEncryptorSupplier());
    serverTransfer.messageDigester = (new SwlSha256MessageDigester());
    serverTransfer.obfuscator = (new SwlBase64Obfuscator());

    let certPair = clientTransfer.generateCertPair("test");
    let serverCert = certPair.server;
    let clientCert = certPair.client;

    let clientSendData = clientTransfer.sendByCert(clientCert,
        ["body:123456", "query:user=admin"]);


    let clientId = "127.0.0.1";
    let serverReceiveData = serverTransfer.receiveByCert(
        clientSendData,
        serverCert,
        clientId);

    let serverResponseData = serverTransfer.sendByCert(serverCert,
        ["echo:ok", "data:ok"]);

    let serverId = "server";
    let clientReceiveData = clientTransfer.receiveByCert(
        serverResponseData,
        clientCert
    );

    console.log("ok");
}

TestSwlExchanger.testRaw=function(){
    let clientTransfer = new SwlExchanger();
    clientTransfer.asymmetricEncryptorSupplier = (new SwlRsaAsymmetricEncryptorSupplier());
    clientTransfer.symmetricEncryptorSupplier = (new SwlAesSymmetricEncryptorSupplier());
    clientTransfer.messageDigester = (new SwlSha256MessageDigester());
    clientTransfer.obfuscator = (new SwlBase64Obfuscator());

    let serverTransfer = new SwlExchanger();
    serverTransfer.asymmetricEncryptorSupplier = (new SwlRsaAsymmetricEncryptorSupplier());
    serverTransfer.symmetricEncryptorSupplier = (new SwlAesSymmetricEncryptorSupplier());
    serverTransfer.messageDigester = (new SwlSha256MessageDigester());
    serverTransfer.obfuscator = (new SwlBase64Obfuscator());


    let serverKeyPair = clientTransfer.generateKeyPair();
    let serverAsymSign = "server";

    let clientKeyPair = clientTransfer.generateKeyPair();
    let clientAsymSign = "client";


    let clientSendData = clientTransfer.sendByRaw(serverKeyPair.getPublicKey(), serverAsymSign,
        clientKeyPair.getPrivateKey(), clientAsymSign,
        ["body:123456", "query:user=admin"]);


    let clientId = "127.0.0.1";
    let serverReceiveData = serverTransfer.receiveByRaw(clientId,
        clientSendData,
        clientKeyPair.getPublicKey(),
        serverKeyPair.getPrivateKey());

    let serverResponseData = serverTransfer.sendByRaw(clientKeyPair.getPublicKey(), clientAsymSign,
        serverKeyPair.getPrivateKey(), serverAsymSign
        , ["echo:ok", "data:ok"]);

    let serverId = "server";
    let clientReceiveData = clientTransfer.receiveByRaw(serverId,
        serverResponseData,
        serverKeyPair.getPublicKey(),
        clientKeyPair.getPrivateKey());

    console.log("ok");
}

export default TestSwlExchanger
